src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,630 @@
+        }
+
+        // Writes the current header as a Gnu entry into the archive stream.
+        // Makes sure to add the preceding LongLink and/or LongPath entries if necessary, before the actual entry.
+        internal void WriteAsGnu(Stream archiveStream, Span<byte> buffer)
+        {
+            // First, we determine if we need a preceding LongLink, and write it if needed
+            if (_linkName.Length > FieldLengths.LinkName)
+            {
+                TarHeader longLinkHeader = GetGnuLongMetadataHeader(TarEntryType.LongLink, _linkName);
+                longLinkHeader.WriteAsGnuInternal(archiveStream, buffer);
+                buffer.Clear(); // Reset it to reuse it
+            }
+
+            // Second, we determine if we need a preceding LongPath, and write it if needed
+            if (_name.Length > FieldLengths.Name)
+            {
+                TarHeader longPathHeader = GetGnuLongMetadataHeader(TarEntryType.LongPath, _name);
+                longPathHeader.WriteAsGnuInternal(archiveStream, buffer);
+                buffer.Clear(); // Reset it to reuse it
+            }
+
+            // Third, we write this header as a normal one
+            WriteAsGnuInternal(archiveStream, buffer);
+        }
+
+        // Creates and returns a GNU long metadata header, with the specified long text written into its data stream.
+        private static TarHeader GetGnuLongMetadataHeader(TarEntryType entryType, string longText)
+        {
+            Debug.Assert((entryType is TarEntryType.LongPath && longText.Length > FieldLengths.Name) ||
+                         (entryType is TarEntryType.LongLink && longText.Length > FieldLengths.LinkName));
+
+            TarHeader longMetadataHeader = default;
+
+            longMetadataHeader._name = GnuLongMetadataName; // Same name for both longpath or longlink
+            longMetadataHeader._mode = (int)TarHelpers.DefaultMode;
+            longMetadataHeader._uid = 0;
+            longMetadataHeader._gid = 0;
+            longMetadataHeader._mTime = DateTimeOffset.MinValue; // 0
+            longMetadataHeader._typeFlag = entryType;
+
+            longMetadataHeader._dataStream = new MemoryStream();
+            longMetadataHeader._dataStream.Write(Encoding.UTF8.GetBytes(longText));
+            longMetadataHeader._dataStream.Seek(0, SeekOrigin.Begin); // Ensure it gets written into the archive from the beginning
+
+            return longMetadataHeader;
+        }
+
+        // Writes the current header as a GNU entry into the archive stream.
+        internal void WriteAsGnuInternal(Stream archiveStream, Span<byte> buffer)
+        {
+            // Unused GNU fields: offset, longnames, unused, sparse struct, isextended and realsize
+            // If this header came from another archive, it will have a value
+            // If it was constructed by the user, it will be an empty array
+            _gnuUnusedBytes ??= new byte[FieldLengths.AllGnuUnused];
```

Comment:
I'm confused by this.  We're never filling this array.  Why is it needed at all?

Reply:
It's a TODO (I added it to TarHeader.Read.cs): If the user generates a GNU tar from another tool, I need to make sure to preserve the data in the header struct, in case the user wants to write this entry to another archive, so we don't lose data.